### PR TITLE
Merge changes to the joint trajectory controller

### PR DESCRIPTION
**Pull request by kbogert**
This commit allows me to build and run this ROS2 package, but I have not tested the changes to the scaled joint trajectory controller (I just merged changes from [https://github.com/ros-controls/ros2_controllers/commit/612f610c24d026a41abd2dd026902c672cf778c9
](https://github.com/ros-controls/ros2_controllers/commit/612f610c24d026a41abd2dd026902c672cf778c9
) )

See also: [https://github.com/ros-controls/ros2_control/commit/156a3f6aaed319585a8a1fd445693e2e08c30ccd](https://github.com/ros-controls/ros2_control/commit/156a3f6aaed319585a8a1fd445693e2e08c30ccd)

### CODE REVIEW

**Comment by schornakj** [comment]
This is an important change to include to allow this package to build against newer versions of `ros2_controllers`. For example, I'd like to use the new [gripper controllers](https://github.com/ros-controls/ros2_controllers/tree/master/gripper_controllers), but the PR adding those was merged in after 0.2.1, which is the version currently used in this package's .repos file.

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 1]
```diff
@@ -55,24 +55,28 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
...
-    }
     return controller_interface::return_type::OK;
   }
 
-  auto resize_joint_trajectory_point = [](trajectory_msgs::msg::JointTrajectoryPoint& point, size_t size) {
+  auto resize_joint_trajectory_point = [&](trajectory_msgs::msg::JointTrajectoryPoint& point, size_t size) {
     point.positions.resize(size);
-    point.velocities.resize(size);
-    point.accelerations.resize(size);
+if (has_velocity_state_interface_) {
```
Comment (destogl) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:63]: Indentation seems to be broken here.
```suggestion
    if (has_velocity_state_interface_) {
```

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 2]
```diff
@@ -89,13 +93,44 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
...
-  }
+  auto assign_point_from_interface = [&, joint_num](
+    std::vector<double> & trajectory_point_interface, const auto & joint_inteface)
+    {
+      for (auto index = 0ul; index < joint_num; ++index) {
+        trajectory_point_interface[index] = joint_inteface[index].get().get_value();
+      }
+    };
+  // TODO(anyone): can I here also use const on joint_interface since the reference_wrapper is not
+  // changed, but its value only?
```
Comment (destogl) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:104]: Removing comment.
```suggestion
```

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 3]
```diff
@@ -89,13 +93,44 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
...
+    auto & joint_inteface, const std::vector<double> & trajectory_point_interface)
+    {
+      for (auto index = 0ul; index < joint_num; ++index) {
+        joint_inteface[index].get().set_value(trajectory_point_interface[index]);
+      }
+    };
+
   state_current.time_from_start.set__sec(0);
 
+// Assign values from the hardware
```
Comment (destogl) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:115]:
```suggestion
  // Assign values from the hardware
```

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 4]
```diff
@@ -123,9 +158,26 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
       bool abort = false;
       bool outside_goal_tolerance = false;
       const bool before_last_point = end_segment_itr != (*traj_point_active_ptr_)->end();
+
```
Comment (destogl) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:161]:
```suggestion
```

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 5]
```diff
@@ -123,9 +158,26 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
...
+        assign_interface_from_point(joint_command_interface_[1], state_desired.velocities);
+      }
+      if (has_acceleration_command_interface_) {
+        assign_interface_from_point(joint_command_interface_[2], state_desired.accelerations);
+      }
+      // TODO(anyone): Add here "if using_closed_loop_hw_interface_adapter" (see ROS1) - #171
+//       if (check_if_interface_type_exist(
+//           command_interface_types_, hardware_interface::HW_IF_EFFORT)) {
+//         assign_interface_from_point(joint_command_interface_[3], state_desired.effort);
+//       }
```
Comment (destogl) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:177]: This is not relevant for the UR.
```suggestion
```

**Review by destogl — CHANGES_REQUESTED** [review]
Please consider comments. We also need someone testing this on hardware than we can merge it. From code/theoretical perspective looks good!

ros-controls/ros2_controllers#189 wil enable some simplifications for this PR, but I would go for merge if we manage to test this on HW soon.

**Comment by destogl** [comment]
Resolves #118

**Comment by AdamPettinger** [comment]
I just tried this. With `use_fake_hardware=false` AND `use_fake_hardware=true` I get and error. Briefly looked and am not sure exactly why (the code looks very close to the `ros2_controllers/JointTrajectoryController`

> [ros2_control_node-1] terminate called after throwing an instance of 'std::runtime_error'
[ros2_control_node-1]   what():  bool joint_trajectory_controller::Trajectory::sample(const rclcpp::Time&, trajectory_msgs::msg::JointTrajectoryPoint&, joint_trajectory_controller::TrajectoryPointConstIter&, joint_trajectory_controller::TrajectoryPointConstIter&) failed. trajectory_msg_ is null.

That is pulling your branch, plus the current `main` from `ros2_controllers`. Launched with:
`ros2 launch ur_bringup ur_control.launch.py ur_type:=ur5e robot_ip:=yyy.yyy.yyy.yyy use_fake_hardware:=true launch_rviz:=false robot_controller:=scaled_joint_trajectory_controller`

**Comment by kbogert** [comment]
In my testing the controller works now with both fake hardware and a real UR3e.  But, MoveIt has some issues with it, in particular the ur_moveit_config/config/controllers.yaml file needs to be changed to use the scaled_joint_trajectory_controller instead of the vanilla joint_trajectory_controller, and the follow joint trajectory timeout needs to be increased or else all trajectories are aborted when the speed scaling is low.

**Comment by AdamPettinger** [comment]
Just tested the new changes on fake and real hardware. Looks great! Scaling worked well, everything ran perfect. Awesome!

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp [hunk 6]
```diff
@@ -55,24 +55,28 @@ controller_interface::return_type ScaledJointTrajectoryController::update()
...
+  auto resize_joint_trajectory_point = [&](trajectory_msgs::msg::JointTrajectoryPoint& point, size_t size) {
     point.positions.resize(size);
-    point.velocities.resize(size);
-    point.accelerations.resize(size);
+    if (has_velocity_state_interface_) {
+        point.velocities.resize(size);
+      }
+      if (has_acceleration_state_interface_) {
+        point.accelerations.resize(size);
+      }
```
Comment (MGBla) [inline ur_controllers/src/scaled_joint_trajectory_controller.cpp:68]: Formating broken. In general please run clang-format over your commit using the provided .clang-format file

**Review by MGBla — CHANGES_REQUESTED** [review]
Hi and thanks for the changes. I just tested it on a UR3e here and everything works as expected. However we still have to satisfy the CI pipeline. This requires only minor stuff:
- [ ] Fix formating of the code using clang-format file
- [ ] Change back to current ros2_controllers version in the [repo](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/Universal_Robots_ROS2_Driver.repos) file
- [ ] add ``` #include <vector> ``` to the header of the scaled_joint_trajectory_controller  

I think that should cover it if I haven't overlooked smth

File: ur_controllers/src/scaled_joint_trajectory_controller.cpp (hunk 6 above)
Comment (AdamPettinger) [reply ur_controllers/src/scaled_joint_trajectory_controller.cpp:68]: @kbogert An easy way to help with CI passing is to use `pre-commit`, check [this section](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver#contributor-guidelines) of the readme. If you install `pre-commit` before you change the repos file and add `#include <vector>`, it'll format it for you when you commit those changes

**Comment by destogl** [comment]
Please check #133 for the follow up on this PR.

**Pull request closed without merging**